### ISSUE TITLE
How to pass onClick to react-map-gl InteractiveMap

### ISSUE THREAD

**Issue by kylebarron** [uber/deck.gl#4093]
I'm thrilled I actually figured this out while writing up the issue. I'll leave it open in case you want to add something to the docs, but otherwise you can close.

----

I'm trying to render a custom Mapbox Vector Tiles source and have the ability to query the rendered features.

For context, my `Source` is selected polygons of western US National Parks, defined by [this Tile JSON](https://tiles.nst.guide/nationalpark/tile.json), and I'm trying to add a simple fill. The vector tiles were made with mapbox/tippecanoe and have embedded properties to query.

## Attempts

I've made three attempts: 

1. pure Deck.gl rendered through `TileLayer`
2. pure React Map GL
3. React Map `Source` as a child of Deck.gl. The third combines Mapbox's rendering of vector tiles with the ability to query from Deck.gl, and is the best in my opinion.

### Pure Deck.gl

The first attempt was to render this through a deck.gl `TileLayer`. After correcting for overlapping features at tile boundaries, the [vector tiles display but have flickering](https://github.com/uber/deck.gl/issues/3824#issuecomment-557831473) when panning or zooming.

This method has the benefit of not needing to interact with `react-map-gl`, but the rendering isn't as good as Mapbox's own rendering.

### Pure React Map GL

The next attempt was a pure React Map GL solution, to make sure that what I want is possible without Deck.GL on top. When I pan/zoom I get the [error in this issue](https://github.com/uber/react-map-gl/issues/978), but otherwise, clicking the polygon of Yosemite correctly logs the Yosemite polygon with all its attributes. This is what I want to achieve through Deck.gl.

Map with Yosemite polygon provided through vector `Source`:
![image](https://user-images.githubusercontent.com/15164633/71755870-071cb180-2e4a-11ea-8862-6e5632fbc227.png)

JSON output of clicked feature:
```json
[{
  "geometry": {
      "type ": "MultiPolygon ",
      "coordinates ": [...]
  },
  "type ": "Feature ",
  "properties ": {
      "OBJECTID ": 261,
      "UNIT_CODE ": "YOSE ",
      "GIS_Notes ": "Lands - http: //landsnet.nps.gov/tractsnet/documents/YOSE/METADATA/yose_metadata.xml",
      "UNIT_NAME": "Yosemite National Park",
      "DATE_EDIT": "2016-01-27T00:00:00.000Z",
      "STATE": "CA",
      "REGION": "PW",
      "GNIS_ID": "255923",
      "UNIT_TYPE": "National Park",
      "CREATED_BY": "Lands",
      "METADATA": "https://irma.nps.gov/App/Reference/Profile/2210713#Yosemite National Park",
      "PARKNAME": "Yosemite",
      "GlobalID": "ad409555-6373-4996-bd50-a6f8a4957cae",
      "Shape__Are": 4849245281.22656,
      "Shape__Len": 379891.899023454,
      "index_right": 0
  },
  "layer": {
      "id": "nationalpark_fill",
      "type": "fill",
      "source": "nationalpark",
      "source-layer": "nationalpark",
      "paint": {
          "fill-opacity": 1,
          "fill-color": "#d8e8c8"
      }
  },
  "source": "nationalpark",
  "sourceLayer": "nationalpark",
  "state": {}
}]
```

Code:
```jsx
import React from "react";
import InteractiveMap, { Source, Layer } from "react-map-gl";
import "./mapbox-gl.css";

const mapStyle =
  "https://raw.githubusercontent.com/nst-guide/osm-liberty-topo/gh-pages/style.json";

class Map extends React.Component {
  state = {
    width: 590,
    height: 946,
    latitude: 37.8759,
    longitude: -119.3629,
    zoom: 8,
    hoveredFeature: null,
    offsetX: null,
    offsetY: null
  };

  _onClick = event => {
    const {
      features,
      srcEvent: { offsetX, offsetY }
    } = event;
    console.log(features)
  };

  render() {
    const { width, height, latitude, longitude, zoom } = this.state;
    return (
      <InteractiveMap
        width={width}
        height={height}
        latitude={latitude}
        longitude={longitude}
        zoom={zoom}
        mapStyle={mapStyle}
        onViewportChange={viewport => {
          const { width, height, latitude, longitude, zoom } = viewport;
          this.setState({
            width: width,
            height: height,
            latitude: latitude,
            longitude: longitude,
            zoom: zoom
          });
        }}
        // interactiveLayerIds={["nationalpark_fill"]}
        onClick={this._onClick}
      >
        <Source
          id="nationalpark"
          type="vector"
          url="https://tiles.nst.guide/nationalpark/tile.json"
        >
          <Layer
            id="nationalpark_fill"
            type="fill"
            source-layer="nationalpark"
            paint={{
              "fill-opacity": 1,
              "fill-color": "#d8e8c8"
            }}
          />
        </Source>
      </InteractiveMap>
    );
  }
}

export default Map;
```

### React Map GL `Source` within Deck.gl

In the midst of writing this issue, I figured out the solution!

When you wrap React Map GL sources with Deck.gl, Deck.gl intercepts all the click/hover events, so you can no longer pass an `onClick` prop to `InteractiveMap`. Instead, when you click on a Deck.gl map, it passes an event object to the `onClick` prop that contains the `x` and `y` values in pixel coordinates that were clicked. You can then pass those `x` and `y` coordinates to [`InteractiveMap.queryRenderedFeatures`](https://uber.github.io/react-map-gl/#/documentation/api-reference/static-map?section=methods) to find the features under that point. You get the same output as through vanilla React Map GL

Working example code:
```jsx
import DeckGL from "@deck.gl/react";
import React from "react";
import InteractiveMap, {
  _MapContext as MapContext,
  Source,
  Layer
} from "react-map-gl";
import "./mapbox-gl.css";

const initialViewState = {
  bearing: 0,
  latitude: 37.8759,
  longitude: -119.3629,
  pitch: 0,
  zoom: 8
};

const mapStyle =
  "https://raw.githubusercontent.com/nst-guide/osm-liberty-topo/gh-pages/style.json";

class Map extends React.Component {
  _onClick = event => {
    // event.x and event.y hold the clicked x and y coordinates in pixels
    // You can pass those coordinates to React Map GL's queryRenderedFeatures
    // to query any desired layers rendered there.
    // Make sure you create the ref on InteractiveMap or StaticMap
    const features = this.map.queryRenderedFeatures([event.x, event.y], {
      layers: ["nationalpark_fill"]
    });
    console.log(features);
  };

  render() {
    return (
      <DeckGL
        initialViewState={initialViewState}
        controller={true}
        ContextProvider={MapContext.Provider}
        onClick={this._onClick}
      >
        <InteractiveMap
          ref={ref => {
            this.map = ref && ref.getMap();
          }}
          mapStyle={mapStyle}
          touchRotate={true}
        >
          <Source
            id="nationalpark"
            type="vector"
            url="https://tiles.nst.guide/nationalpark/tile.json"
          >
            <Layer
              id="nationalpark_fill"
              type="fill"
              source-layer="nationalpark"
              paint={{
                "fill-opacity": 1,
                "fill-color": "#d8e8c8"
              }}
            />
          </Source>
        </InteractiveMap>
      </DeckGL>
    );
  }
}

export default Map;
```

Example feature output
```json
[{
    "geometry": {
        "type": "Polygon",
        "coordinates": [...]
    },
    "type": "Feature",
    "properties": {
        "OBJECTID": 261,
        "UNIT_CODE": "YOSE",
        "GIS_Notes": "Lands - http://landsnet.nps.gov/tractsnet/documents/YOSE/METADATA/yose_metadata.xml",
        "UNIT_NAME": "Yosemite National Park",
        "DATE_EDIT": "2016-01-27T00:00:00.000Z",
        "STATE": "CA",
        "REGION": "PW",
        "GNIS_ID": "255923",
        "UNIT_TYPE": "National Park",
        "CREATED_BY": "Lands",
        "METADATA": "https://irma.nps.gov/App/Reference/Profile/2210713#Yosemite National Park",
        "PARKNAME": "Yosemite",
        "GlobalID": "ad409555-6373-4996-bd50-a6f8a4957cae",
        "Shape__Are": 4849245281.22656,
        "Shape__Len": 379891.899023454,
        "index_right": 0
    },
    "layer": {
        "id": "nationalpark_fill",
        "type": "fill",
        "source": "nationalpark",
        "source-layer": "nationalpark",
        "paint": {
            "fill-opacity": 1,
            "fill-color": "#d8e8c8"
        }
    },
    "source": "nationalpark",
    "sourceLayer": "nationalpark",
    "state": {}
}]
```

Ref #3824 , #3389

**Comment by Pessimistress** [comment]
Just to clarify, it is recommended to use `<StaticMap>` as `<DeckGL>`'s child. Your example does not use any interactivity from `<InteractiveMap>` - Deck is handling all the user inputs here. `<StaticMap>` provides considerably better performance.

**Comment by kylebarron** [comment]
I'm not sure exactly why, but replacing `InteractiveMap` with `StaticMap` raises an error on rendering the map. 

Example code, the third (correct) example with just `InteractiveMap` -> `StaticMap`:
```jsx
import DeckGL from "@deck.gl/react";
import React from "react";
import {
  StaticMap,
  _MapContext as MapContext,
  Source,
  Layer
} from "react-map-gl";
import "./mapbox-gl.css";

const initialViewState = {
  bearing: 0,
  latitude: 37.8759,
  longitude: -119.3629,
  pitch: 0,
  zoom: 8
};

const mapStyle =
  "https://raw.githubusercontent.com/nst-guide/osm-liberty-topo/gh-pages/style.json";

class Map extends React.Component {
  _onClick = event => {
    // event.x and event.y hold the clicked x and y coordinates in pixels
    // You can pass those coordinates to React Map GL's queryRenderedFeatures
    // to query any desired layers rendered there.
    // Make sure you create the ref on InteractiveMap or StaticMap
    const features = this.map.queryRenderedFeatures([event.x, event.y], {
      layers: ["nationalpark_fill"]
    });
    console.log(features);
  };

  render() {
    return (
      <DeckGL
        initialViewState={initialViewState}
        controller={true}
        ContextProvider={MapContext.Provider}
        onClick={this._onClick}
      >
        <StaticMap
          ref={ref => {
            this.map = ref && ref.getMap();
          }}
          mapStyle={mapStyle}
        >
          <Source
            id="nationalpark"
            type="vector"
            url="https://tiles.nst.guide/nationalpark/tile.json"
          >
            <Layer
              id="nationalpark_fill"
              type="fill"
              source-layer="nationalpark"
              paint={{
                "fill-opacity": 1,
                "fill-color": "#d8e8c8"
              }}
            />
          </Source>
        </StaticMap>
      </DeckGL>
    );
  }
}

export default Map;
```

Traceback:
```
Uncaught TypeError: Cannot read property 'on' of null
    at Source._render (source.js:133)
    at updateContextConsumer (react-dom.development.js:19844)
    at beginWork$1 (react-dom.development.js:20227)
    at HTMLUnknownElement.callCallback (react-dom.development.js:336)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:385)
    at invokeGuardedCallback (react-dom.development.js:440)
    at beginWork$$1 (react-dom.development.js:25780)
    at performUnitOfWork (react-dom.development.js:24695)
    at workLoopSync (react-dom.development.js:24671)
    at performSyncWorkOnRoot (react-dom.development.js:24270)
    at react-dom.development.js:12199
    at unstable_runWithPriority (scheduler.development.js:697)
    at runWithPriority$2 (react-dom.development.js:12149)
    at flushSyncCallbackQueueImpl (react-dom.development.js:12194)
    at flushSyncCallbackQueue (react-dom.development.js:12182)
    at scheduleUpdateOnFiber (react-dom.development.js:23709)
    at Object.enqueueForceUpdate (react-dom.development.js:14033)
    at DeckGL.push../node_modules/react/cjs/react.development.js.Component.forceUpdate (react.development.js:344)
    at DeckGL._customRender (deckgl.js:151)
    at Deck.redraw (deck.js:351)
    at Deck._onRenderFrame (deck.js:707)
    at AnimationLoop.onRender (animation-loop.js:252)
    at AnimationLoop._renderFrame (animation-loop.js:340)
    at AnimationLoop.redraw (animation-loop.js:178)
    at renderFrame (animation-loop.js:279)
index.js:1 The above error occurred in the <Context.Consumer> component:
    in Source (at App.js:49)
    in div (created by Context.Consumer)
    in div (created by AutoSizer)
    in AutoSizer (created by StaticMap)
    in div (created by StaticMap)
    in StaticMap (at App.js:42)
    in div (created by DeckGL)
    in div (created by DeckGL)
    in DeckGL (at App.js:36)
    in Map (at src/index.js:7)

Consider adding an error boundary to your tree to customize error handling behavior.
Visit https://fb.me/react-error-boundaries to learn more about error boundaries.
console.<computed> @ index.js:1
react-dom.development.js:12212 Uncaught TypeError: Cannot read property 'on' of null
    at Source._render (source.js:133)
    at updateContextConsumer (react-dom.development.js:19844)
    at beginWork$1 (react-dom.development.js:20227)
    at HTMLUnknownElement.callCallback (react-dom.development.js:336)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:385)
    at invokeGuardedCallback (react-dom.development.js:440)
    at beginWork$$1 (react-dom.development.js:25780)
    at performUnitOfWork (react-dom.development.js:24695)
    at workLoopSync (react-dom.development.js:24671)
    at performSyncWorkOnRoot (react-dom.development.js:24270)
    at react-dom.development.js:12199
    at unstable_runWithPriority (scheduler.development.js:697)
    at runWithPriority$2 (react-dom.development.js:12149)
    at flushSyncCallbackQueueImpl (react-dom.development.js:12194)
    at flushSyncCallbackQueue (react-dom.development.js:12182)
    at scheduleUpdateOnFiber (react-dom.development.js:23709)
    at Object.enqueueForceUpdate (react-dom.development.js:14033)
    at DeckGL.push../node_modules/react/cjs/react.development.js.Component.forceUpdate (react.development.js:344)
    at DeckGL._customRender (deckgl.js:151)
    at Deck.redraw (deck.js:351)
    at Deck._onRenderFrame (deck.js:707)
    at AnimationLoop.onRender (animation-loop.js:252)
    at AnimationLoop._renderFrame (animation-loop.js:340)
    at AnimationLoop.redraw (animation-loop.js:178)
    at renderFrame (animation-loop.js:279)
```

**Comment by Pessimistress** [comment]
Hmm this is not really a scenario we tested for. It should theoretically go away if you drop the `ContextProvider` prop.

**Comment by kylebarron** [comment]
I think I'm using `ContextProvider` to show the `NavigationControl` and `ScaleControl`

**Comment by Pessimistress** [comment]
@kylebarron Just published `react-map-gl@5.2.3` that should solve the crash. If it's still an issue let's move this conversation to the react-map-gl repo.

**Comment by kylebarron** [comment]
It does appear to solve the crash. Thanks!
